pkg/storage/statestore.go

```diff
@@ -6,21 +6,161 @@ package storage
 
 import (
 	"io"
+	"path"
 
-	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/vmihailenco/msgpack/v5"
 )
 
-// StateStorer defines methods required to get, set, delete values for different keys
-// and close the underlying resources.
+// StateIterFunc is used when iterating through StateStorer key/value pairs
+type StateIterFunc func(key, val []byte) (stop bool, err error)
+
+// StateStorer is a storage interface for storing and retrieving key/value pairs.
 type StateStorer interface {
-	Get(key string, i interface{}) (err error)
-	Put(key string, i interface{}) (err error)
-	Delete(key string) (err error)
-	Iterate(prefix string, iterFunc StateIterFunc) (err error)
-	// DB returns the underlying DB storage.
-	DB() *leveldb.DB
 	io.Closer
+
+	// Get unmarshalls object with the given key into the given obj.
+	Get(key string, obj interface{}) error
+
+	// Put inserts or updates the given obj stored under the given key.
+	Put(key string, obj interface{}) error
+
+	// Delete removes object form the store stored under the given key.
+	Delete(key string) error
+
+	// Iterate iterates over all keys with the given prefix and calls iterFunc.
+	Iterate(prefix string, iterFunc StateIterFunc) error
+
+	// DB returns the underlying DB storage.
+	//
+	// Note: the returned interface is a gross hack until
+	// we can refactor the kademlia to not use the shed.
+	DB() interface{}
 }
 
-// StateIterFunc is used when iterating through StateStorer key/value pairs
-type StateIterFunc func(key, value []byte) (stop bool, err error)
+// stateStoreNamespace is the namespace used for state storage.
+const stateStoreNamespace = "statestore"
+
+var _ Item = (*proxyItem)(nil)
+
+// proxyItem is a proxy object that implements the Item interface.
+// It is an intermediary object between StateStorer and Store interfaces calls.
+type proxyItem struct {
+	ns  string
+	key string
+	obj interface{}
+}
+
+// ID implements Item interface.
+func (ip *proxyItem) ID() string {
+	return ip.key
+}
+
+// Namespace implements Item interface.
+func (ip *proxyItem) Namespace() string {
+	return ip.ns
+}
+
+// Marshal implements Item interface.
+func (ip *proxyItem) Marshal() (data []byte, err error) {
+	return msgpack.Marshal(ip.obj)
```

Comment:
I think in the marshal we should check
`bObj, ok := ip.obj.(encoding.BinaryMarshaler)`

If it implements it, we use `bObj.Marshal()`. Same for the unmarshal part.

Reply:
Done.